lib/galaxy/tools/toolbox/lineages/factory.py

```diff
@@ -20,28 +11,39 @@ def __init__(self, app):
         self.lineage_map = {}
         self.app = app
 
-    def register(self, tool, from_toolshed=False):
+    def register(self, tool):
         tool_id = tool.id
-        versionless_tool_id = remove_version_from_guid( tool_id )
-        if from_toolshed:
-            lineage = ToolShedLineage.from_tool(self.app, tool)
-        else:
+        versionless_tool_id = remove_version_from_guid( tool_id)
+        lineage = self.lineage_map.get(versionless_tool_id)
+        if not lineage:
             lineage = StockLineage.from_tool( tool )
         if versionless_tool_id and versionless_tool_id not in self.lineage_map:
             self.lineage_map[versionless_tool_id] = lineage
         if tool_id not in self.lineage_map:
             self.lineage_map[tool_id] = lineage
+        lineage.register_version(tool.version)
```

Comment:
At first glance I agree with you, but removing this line breaks the lineages ... I'll have to dig into why that happens